### PR TITLE
GDPR PII/OII Filtering for Logger

### PR DESCRIPTION
**Pull request by heidijinxujia**
Ref #1022 
- Rewrite the Logger class
- Add GDPR PII/OII Filtering for Logger
- Put all the PII info into additional message in the Logger callings.
- Fix PMD and checkstyle errors.
- Fix the unit tests

### CODE REVIEW

**Comment by heidijinxujia** [comment]
To-do:
Need to check the sov_cloud change on GDPR.

File: adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java [hunk 1]
```diff
@@ -29,6 +29,7 @@
 import android.os.Handler;
 import android.os.HandlerThread;
 import android.support.annotation.Nullable;
+import android.util.Log;
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java]: why do you need this?

File: adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java [hunk 2]
```diff
@@ -434,8 +439,8 @@ private void logReturnedToken(final AuthenticationResult result) {
         if (result != null && result.getAccessToken() != null) {
             String accessTokenHash = getTokenHash(result.getAccessToken());
             String refreshTokenHash = getTokenHash(result.getRefreshToken());
-            Logger.v(TAG, String.format(
-                    "Access TokenID %s and Refresh TokenID %s returned.",
+            Logger.i(TAG, "Access tokenID and refresh tokenID returned. ", String.format(
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java]: delete this function, we should not log tokens

File: adal/src/main/java/com/microsoft/aad/adal/Logger.java [hunk 3]
```diff
@@ -154,163 +194,59 @@ private static String addMoreInfo(String message) {
...
+        final StringBuilder logMessage = new StringBuilder();
 
-    /**
-     *
-     * @param tag tag for the log message
-     * @param message body of the log message
-     * @param additionalMessage additional parameters
-     * @param errorCode ADAL error code being logged
-     */
-    public void inform(String tag, String message, String additionalMessage, ADALError errorCode) {
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/Logger.java]: was all this dead code?

File: adal/src/main/java/com/microsoft/aad/adal/Logger.java [hunk 4]
```diff
@@ -154,163 +194,59 @@ private static String addMoreInfo(String message) {
...
-        if (mLogLevel.compareTo(LogLevel.Warn) < 0) {
-            return;
+        if (message != null) {
+            logMessage.append(' ').append(addMoreInfo(message));
         }
 
-        if (mAndroidLogEnabled) {
-            Log.w(tag, getLogMessage(message, additionalMessage, errorCode));
+        // Adding stacktrace to message
+        if (throwable != null) {
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/Logger.java]: shouldn't you add a PII check around this?

Comment (heidijinxujia) [reply adal/src/main/java/com/microsoft/aad/adal/Logger.java]: The original idea is that we only use throwable here when it is no a PII exception, and if the PII exception is thrown, we add the stacktrace into the additional msg, where we do the PII flag check. Is there anything I misunderstood here?

File: adal/src/main/java/com/microsoft/aad/adal/Logger.java (hunk 3 above)
Comment (heidijinxujia) [reply adal/src/main/java/com/microsoft/aad/adal/Logger.java]: Yes, they are all dead now. I replace all these functions with loglevel parameter of function log().

File: adal/src/main/java/com/microsoft/aad/adal/Logger.java [hunk 5]
```diff
@@ -30,22 +30,75 @@
...
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * Android log output can. If externalLogger is set, it will use that as well.
- * Usage: Logger.v(TAG, message, additionalMessage, errorCode) to log. Set
+ * Usage: Logger.v(tag, message, additionalMessage, errorCode) to log. Set
  * custom logger: Logger.setExternalLogger(..);
  */
 public class Logger {
+    private static Logger INSTANCE = new Logger();
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/Logger.java]: FindBugs is throwing a rules violation on this name
```
warning: Name 'INSTANCE' must match pattern '^[s][a-zA-Z0-9]*$'.
```

**Comment by heidijinxujia** [comment]
@iambmelt Thank you for the comment. All the unit test fixed.

File: adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java [hunk 6]
```diff
@@ -777,12 +793,15 @@ public void run() {
...
+                                            authenticationException.getCode() == null ? ADALError.AUTHORIZATION_CODE_NOT_EXCHANGED_FOR_TOKEN.getDescription() : authenticationException.getCode().getDescription(),
+                                            message.toString()
+                                                    + ' ' + ExceptionExtensions.getExceptionMessage(authenticationException)
+                                                    + ' ' + Log.getStackTraceString(authenticationException),
                                             ADALError.AUTHORIZATION_CODE_NOT_EXCHANGED_FOR_TOKEN,
-                                            authenticationException);
+                                            null);
                                     waitingRequestOnError(callbackHandle, waitingRequest, requestId,
-                                            authenticationException);
+                                            null);
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java]: Was this change intentional?
